### ISSUE TITLE
Question about to get gt label of custom dataset without depth file.

### ISSUE THREAD

**Issue by Irelia-go** [zju3dv/LoFTR#324]
Thanks for the author's great work, I got good performance in training with Megadepth. However, I also want to train LOFTR with custom dataset which has pair images and matching points without depth file. I have noticed some suggestions at #254 #243 and so on, and try to build necessary gt labels in released code. According to updating keys in fucntion "spvs_coarse", these are necessary:

1. conf_matrix_gt
2. spv_b_ids, spv_i_ids, spv_j_ids
3. spv_pt1_i
4. spv_w_pt0_i

Here is a concrete example：

```
batch_size = 1
img0.shape (h,w) [1,1,16,24]
img1.shape (h,w):[1,1,16,24]
scale0(x,y): [1,1]
scale1(x,y):[1,1]
coarse resolution:8
fine resulotion:2
mathing points (x,y): locs0: [[1,1]], locs1: [[8,8]]
```

geting gt 1,2,3 is clearly and easy:
```
1. conf_matrix_gt.shape: [1,(16//8)*(24//8),(16//8)*(24//8)]
    conf_matrix_gt = torch.zeros([1,6,6])
    conf_matrix_gt[0,1//8+1//8*3,8//8+8//8*3]=conf_matrix_gt[0,0,4]=1

2. spv b_id = [[0]]
    spv_i_ids[[0]]
    spv_j_ids = [[4]]

3. use function "create_meshgrid" as wirited in supervision.py
    spv_pt1_i =[[[0,0],[8,0],[16,0],[24,0],...,[24,16]]]
```

I have doubts about how to get spv_w_pt0_i in 4, we can find that there exists scale operation from "corret_mask" in compute_fine_loss and heat_map in fine_matching.
```
4. utilize window size of 5 in coare to fine
    feat_f.shape= [1,128,8,12]
    feat_f = F.unfold(feat_f,kernel_size=(5,5),stride=4,padding=2).reshape[1,6,25,128]
    after getting right prediciton b_ids,i_ids and j_ids in coarse mathing,
    feat_f0.shape=feat_f1.shape=[1,5*5,128]
```
![Image](https://github.com/user-attachments/assets/2b980b73-10b7-4315-9c42-6ff276858c89)
The image shows how to get feat_f by F.unfold. padding=2 ,stride=4,kernel_size=(5,5).The red box indicates the feature area selected by unfold each time, and the yellow dot indicates its center. The yellow point coordinates correspond to the actual coordinates (0, 0), (8, 0), (0, 8), (8, 8), which are consistent with spv_pt1_i. 
```
use function "create_meshgrid" as wirited in supervision.py
spv_w_pt0_i.shape = spv_pt1_i.shape 
spv_w_pt0_i [:,:,:] = float("-inf")
```

(1,1) is in (0.5,0.5) in fine grid and (0.5,0.5) in window and (0,0) in coarse grid,
(8,8) is in (4,4) in fine grid and (0,0) in window and (1,1) in coarse grid
To satisfy scale operation, loc of image 0 should warp to loc of image 1 window , so the final gt `spv_w_pt0_i[0,0] = [4.5,4.5]`
However, when locs like (4,4), there is an overlap between two window feature. I don't known how to correctly process it . Looking forward to your reply and other discussions.

**Comment by HaoyuYao** [comment]
Hello, have addressed the problem? Looking forward to your reply.

**Comment by Irelia-go** [comment]
> Hello, have addressed the problem? Looking forward to your reply.

A little supplement to the above text. Assuming that the matching images only have displacement in horizontal and vertical directions, as long as there is a set value, it can be known which points correspond to each other. This situation represents the simplest case of obtaining labels without depth files.I constructed such a simple dataset on landscape images for training and achieved good results.

**Comment by xyty445671259** [comment]
hi~1.My dataset involves matching two 2D images. Since I don't have pose, depth, or intrinsic parameters, I made some annotations on the two images and used these annotated points to construct a homography matrix for supervision. However, the practical results are not very good, possibly due to my limited understanding of the code and methodology. I would like to ask for suggestions on how to improve this.

For example, dividing the images into a 3x3 grid: most patches correspond one-to-one between the two grids, but some patches may have deviations. For instance, patch 5 in image A might correspond to patch 9 in image B. In such cases, the constructed homography matrix could have significant errors. My current idea is to remove the matching in the fine stage. Is this feasible? Additionally, reducing the downsampling rate—for example, changing Megadepth's 8 to 4—might help the model capture the overall structure and achieve rough matching.

2.Alternatively, are there methods to directly use annotated points for training? Is this feasible? Are there any references or methods available? I am new to matching and would appreciate some guidance. Thank you!

**Comment by Irelia-go** [comment]
> hi~1.My dataset involves matching two 2D images. Since I don't have pose, depth, or intrinsic parameters, I made some annotations on the two images and used these annotated points to construct a homography matrix for supervision. However, the practical results are not very good, possibly due to my limited understanding of the code and methodology. I would like to ask for suggestions on how to improve this.
> 
> For example, dividing the images into a 3x3 grid: most patches correspond one-to-one between the two grids, but some patches may have deviations. For instance, patch 5 in image A might correspond to patch 9 in image B. In such cases, the constructed homography matrix could have significant errors. My current idea is to remove the matching in the fine stage. Is this feasible? Additionally, reducing the downsampling rate—for example, changing Megadepth's 8 to 4—might help the model capture the overall structure and achieve rough matching.
> 
> 2.Alternatively, are there methods to directly use annotated points for training? Is this feasible? Are there any references or methods available? I am new to matching and would appreciate some guidance. Thank you!

I offer some suggestions based on my understanding, for reference only.
1. During the coarse matching stage, the constructed matching labels do not need to be changed. A0 and B0 are the actual matching patch. Even if B1 is very similar to A0 and is matched with A0, it will not affect the matching of other patches. The feature processing of trainsformer should be sufficient.
2. If GPU and inference speed meets the requirements, reducing the coarse mathing path size is theoretically better. Even more exaggerated, a 1*1 block size can be used, thus do not need fine-matching stage.
3. I'm not sure about your dataset format, there is a matrix transformation of the pair images？ The key point is to construct the gt_labels during the coarse and fine matching stages as mentioned before.

**Comment by xyty445671259** [comment]
thank u! for 3, I used labelmatch to mark 20 points between two matching images. Using these 20 points, I constructed the correspondence between the images, which is the transformation matrix. The loss function was changed to projection error. However, this method still does not work well in some scenes. I hope to be able to clearly indicate which points in a pair of matching images are matched. Is this possible? Thanks again






廷玉
***@***.***

**Comment by Sapphire266366** [comment]
> thank u! for 3, I used labelmatch to mark 20 points between two matching images. Using these 20 points, I constructed the correspondence between the images, which is the transformation matrix. The loss function was changed to projection error. However, this method still does not work well in some scenes. I hope to be able to clearly indicate which points in a pair of matching images are matched. Is this possible? Thanks again
> 
> 
> 
> 
> 
> 
> 廷玉
> ***@***.***

Hello, have you solved it? I have been conducting matching experiments recently. Without depth information, currently we can only align the projections of images by manually marking points. Have you used this method to solve the problem yet? Thank you again.
